server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataVerifier.java

```diff
@@ -109,6 +110,14 @@ public IndexMetadata verifyIndexMetadata(IndexMetadata indexMetadata, IndexVersi
      * previous major version.
      */
     private static void checkSupportedVersion(IndexMetadata indexMetadata, IndexVersion minimumIndexCompatibilityVersion) {
+        // Special case for archive indices that have been restored first in a previous major version. Their compatibility version
+        // is < N - 1 and needs special treatment. This can be removed once N - 2 support has been implemented for all indices.
+        if (indexMetadata.getCreationVersion().equals(indexMetadata.getCompatibilityVersion()) == false
+            && indexMetadata.getCreationVersion().isLegacyIndexVersion()
+            && indexMetadata.getCompatibilityVersion().onOrAfter(IndexVersions.MINIMUM_READONLY_COMPATIBLE)) {
+            return;
+        }
```

Comment:
@tlrx are you good with this special case here and the one above that targets archive indices? We are addressing the need to upgrade the compatibility version for archive indices imported first in 8.x, that don't allow to upgrade to 9.x otherwise.

Reply:
I don't think this is going to work during rolling upgrades, because the master may still be a 8.x node and therefore will reject join requests from 9.x, because a 8.x master will still think that 9.x nodes cannot support indices with a 7.x compatibility version.

I think it does work in the case of a full cluster restart though, since the node will be directly be 9.x with the new logic here.

We have to change that logic to also support searchable snapshot N-2 indices, maybe I can do it for both types of indices at once.

Reply:
I'm about to merge https://github.com/elastic/elasticsearch/pull/118744 that is a prerequisite of #118785 which contains similar logic than the one added here but for searchable snapshots.

Reply:
good point Tanguy, thanks for raising this. I guess we are better off waiting for your changes then.

Reply:
I replicated the same test under the `qa/rolling-upgrade` project to verify the behavior. It fails in the line, 
```
NodeJoinExecutor.ensureIndexCompatibility(NodeJoinExecutor.java:386)
```
when upgrading any of the nodes.

Reply:
thanks for doing that!

Reply:
I also got stuck in this area (IndexMetadataVerifier#checkSupportedVersion) when trying to revert one of the deleted FullClusterRestartIT tests. My quick fix of changing IndexVersions.MINIMUM_COMPATIBLE to IndexVersions.MINIMUM_READ_ONLY_COMPATIBLE in https://github.com/elastic/elasticsearch/blob/5ec0a843e186cf7f2a8f1d6120e28f0db4662e38/server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java#L298 led to other startup issues on full cluster restart, I haven't succeeded in exactly debugging those yet but will continue to investigate.

Reply:
Sounds like we may need distrib help with some of these issues @cbuescher , or at least some of the changes they are working on.

Reply:
I opened #118941 (with #118923 for `8.18`) that should allow archive indices in 7.x to exist on a 9.x cluster. I will try to test the change for archive too but in the case you're interested [here is the change](https://github.com/elastic/elasticsearch/pull/118941/files#diff-2a7d8da73e3d928e365dc01640479120b6ef1b06d0b908bd3efcf4f3f6018990R158) for IndexMetadataVerifier.